### ISSUE TITLE
Support SparseLinearConstraint

### ISSUE THREAD

**Issue by hongkai-dai** [RobotLocomotion/drake#10240]
Currently we only support dense matrix `A` in LinearConstraint `lower <= A*x <= upper`. This is both memory inefficient, and slow when constructing/copying a dense matrix. We should support a SparseLinearConstraint as
```C++
class SparseLinearConstraint : public Constraint {
  private:
  Eigen::SparseMatrix<double> A_;
}
```

**Comment by jwnimmer-tri** [comment]
This seems like a sub-issue of #4870.  Marking it as such.

**Comment by jwnimmer-tri** [comment]
This was fixed in #16894.
